Source/Mock.cs

```diff
@@ -838,44 +837,47 @@ private static void SetupAllProperties(Mock mock, Stack<Type> mockedTypesStack)
 			{
 				var expression = GetPropertyExpression(mockType, property);
 				var getter = property.GetGetMethod(true);
-				object value = SetupAllProperties_GetInitialValue(mock, mockedTypesStack, getter);
 
-				var mocked = value as IMocked;
-				if (mocked != null)
+				object value = null;
+				bool valueNotSet = true;
+
+				mock.Setups.Add(new PropertyGetterMethodCall(mock, expression, getter, () =>
 				{
-					SetupAllProperties(mocked.Mock, mockedTypesStack);
-				}
+					if (valueNotSet)
+					{
+						object initialValue;
+						try
+						{
+							initialValue = mock.GetDefaultValue(getter, useAlternateProvider: defaultValueProvider);
+						}
+						catch
+						{
+							// Since this method performs a batch operation, a single failure of the default value
+							// provider should not tear down the whole operation. The empty default value provider
+							// is a safe fallback because it does not throw.
+							initialValue = mock.GetDefaultValue(getter, useAlternateProvider: DefaultValueProvider.Empty);
+						}
+
+						if (initialValue is IMocked mocked)
+						{
+							SetupAllPropertiesPexProtected(mocked.Mock, defaultValueProvider);
+						}
 
-				mock.Setups.Add(new PropertyGetterMethodCall(mock, expression, getter, () => value));
+						value = initialValue;
+						valueNotSet = false;
```

Comment:
In theory, there's room for a race condition here. The uninitialized property could be queried simultaneously several times, or it could be queried while the setter is running, too. In practice, this might not matter much with regards to correctness, but it *might* cause some inefficiency because multiple initial values are produced when only one is required.